engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/EqualsNode.java

```diff
@@ -40,367 +37,184 @@
-   * cannot be present in Enso. All the primitive types should be handled in their corresponding
-   * specializations. See {@link
-   * org.enso.interpreter.node.expression.builtin.interop.syntax.HostValueToEnsoNode}.
-   */
-  static boolean isPrimitive(Object object, InteropLibrary interop) {
-    return isPrimitiveValue(object)
-        || object instanceof EnsoBigInteger
-        || object instanceof Text
-        || interop.isString(object)
-        || interop.isNumber(object)
-        || interop.isBoolean(object);
-  }
 
-  static boolean isPrimitiveValue(Object object) {
-    return object instanceof Boolean || object instanceof Long || object instanceof Double;
-  }
-
-  static boolean isBigInteger(InteropLibrary iop, Object v) {
-    if (v instanceof EnsoBigInteger) {
-      return true;
-    } else {
-      return !iop.fitsInDouble(v) && !iop.fitsInLong(v) && iop.fitsInBigInteger(v);
+    private boolean doDispatch(
+        VirtualFrame frame,
+        Object self,
+        Object that,
+        Type thatType,
+        InteropConversionCallNode convertNode,
+        EqualsSimpleNode equalityNode)
+        throws PanicException {
+      var convert = UnresolvedConversion.build(thatType.getDefinitionScope());
+
+      var ctx = EnsoContext.get(this);
+      var state = State.create(ctx);
+      try {
+        var selfAsThat = convertNode.execute(convert, state, new Object[] {thatType, self});
+        var result = equalityNode.execute(frame, selfAsThat, that);
+        assert !result || assertHashCodeIsTheSame(self, selfAsThat);
+        return result;
+      } catch (ArityException ex) {
+        var assertsOn = false;
+        assert assertsOn = true;
+        if (assertsOn) {
+          throw new AssertionError("Unexpected arity exception", ex);
+        }
+        return false;
+      } catch (PanicException ex) {
+        if (ctx.getBuiltins().error().isNoSuchConversionError(ex.getPayload())) {
+          return false;
+        }
+        throw ex;
+      }
     }
-  }
 
-  BigInteger asBigInteger(InteropLibrary iop, Object v) {
-    if (v instanceof EnsoBigInteger big) {
-      return big.getValue();
-    } else {
-      try {
-        return iop.asBigInteger(v);
-      } catch (UnsupportedMessageException ex) {
-        throw EnsoContext.get(this).raiseAssertionPanic(this, "Expecting BigInteger", ex);
+    private boolean assertHashCodeIsTheSame(Object self, Object converted) {
+      var selfHash = HashCodeNode.getUncached().execute(self);
+      var convertedHash = HashCodeNode.getUncached().execute(converted);
+      var ok = selfHash == convertedHash;
+      var msg =
+          "Different hash code! Original "
+              + self
+              + "[#"
+              + Long.toHexString(selfHash)
+              + "] got converted to "
+              + converted
+              + "[#"
+              + Long.toHexString(convertedHash)
+              + "]";
+      if (!ok) {
+        var ctx = EnsoContext.get(this);
```

Comment:
As a result of this `-ea` only check, I had to update `Complex_Comparator` to return the same hash for numbers without imaginary part as the real part has. I also had to add comparator+hash code for `Fool`: ecd4b90

Reply:
### Flaw in the Design

I am afraid there is a flaw in the design. Once we define `Text.from (that:Any)  = that.to_text`, do we really want
```ruby
obj == obj.to_text
```
? As that's what we would get if we just try to apply conversion to match both sides of `==`.

f33faad mitigates the problem a bit by doing the conversion only when original and converted object have the **same hash code**. That fixes the most apparent problem, but still one could find some `Number` and a `Text` that accidentally have the same hashcode when number is `.to_text` converted...

### TL;DR

Keeping consistency of `==` and `hash` is important for our `Map` implementation. Doing these conversions is important for `Number` and `Decimal` equality. Doing these conversions for any `.from` conversion may yield surprising results.  

I am not sure what to do with that?

Reply:
### Explicitly Define Both Comparators

To guarantee that the implementation of `==` is supposed to use `from` conversion we need the `Complex` (or `Rational` or `Decimal`) module to explicitly state that it supports equality for `Number` (or `Integer`). To do so, let's request the module to define **conversions of both types** to `Comparator`. E.g.:
```ruby
Comparable.from (_:Complex) = Complex_Comparator
Comparable.from (_:Number) = Complex_Comparator
```
Only when **both conversions are defined** and yield the same comparator (in the scope of the module), then `==` functionality will perform the conversion.

##### Rationale

By explicitly specifying multiple `Comparable.from` conversions, the module without any doubt declares its willingness to support the `==` and `hash` code operations for combination of such types. Detecting presence of such conversions is _sufficient check_ for performing the conversions. Conversions defined for other purpose `Text.from (that:Complex)` aren't going to be (accidentally) taken into account without appropriate `Comparable.from (_:Text)` definition.

Such **conversions are local**, however. Other modules like `Map` or `Set` will continue to use the original definition of `Comparable.from (_:Number)` coming from the `Numbers` module. The infrastructure will check for [consistency of hash](https://github.com/enso-org/enso/commit/ecd4b90b834766f33a27d6d721ab1d1173feaa80#diff-b4b39451e4a4157a026bd98878cf0a23c55bbc5245a37a21efcf2de676f59738R199) and yield an `AssertionError` when the `hash` consistency is violated.

Reply:
Initial implementation is in a5c0674c3b